### ISSUE TITLE
deprecate @*, %*, &*, ** too

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12814]
Migrated from [rt.perl.org#116943](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116943) (status was 'rejected')

Searchable as RT116943$

**Comment by p5pRT** [comment]
### From @nwc10

perlvar says&#8203;:

&nbsp;    C\<$\*> was a variable that you could use to enable multiline matching\.
&nbsp;    After a deprecation cycle\, its magic was removed in Perl v5\.10\.0\.
&nbsp;    Using it now triggers a warning&#8203;: C\<$\* is no longer supported>\.
&nbsp;    You should use the C\</s> and C\</m> regexp modifiers instead\.

This is true\. In that currently a warning is issued&#8203;:

$ \./perl \-e '$a = $\*'
$\* is no longer supported at \-e line 1\.

However\, there's no such warning for other things in that typeglob\.

$ \./perl \-Ilib \-e '@&#8203;\* = \( "a"\, "b" \); print "\\$\*\[1\] is $\*\[1\]\\n"'
$\*\[1\] is b
$ \./perl \-Ilib \-e '%\* = \( "a"\, "b" \); print "\\$\*\{a\} is $\*\{a\}\\n"'
$\*\{a\} is b
$ \./perl \-Ilib \-e '\*\* = sub \{ print "Hello world\\n" \}; &\*'
Hello world


I think that we should deprecate the use of \*all\* of these\.
With the intent of freeing up the parser so that things like

&nbsp;   @&#8203;\*array

can become meaningful in the near future\. \(Slurpy arrays and slurpy hashes
would be particularly useful for signatures\)


I've tried to use csearch to find instances of use of these constructions on
CPAN\, but it's very hard to avoid getting a heck of a lot of output from
unrelated things\. As best I could work out \(using a more restricted search
term\)\, they aren't used\. I think that the biggest user may well be JAPHs and
obfuscations relying on this&#8203;:

$ \./perl \-Ilib \-le 'print \*\*\*\*\*\*'
1

Nicholas Clark

**Comment by p5pRT** [comment]
### From @tux

On Tue\, 26 Feb 2013 05&#8203;:41&#8203;:35 \-0800\, Nicholas Clark \(via RT\)
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> \# New Ticket Created by  Nicholas Clark 
> \# Please include the string&#8203;:  \[perl \#116943\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116943 >
> 
> 
> perlvar says&#8203;:
> 
>     C\<$\*> was a variable that you could use to enable multiline matching\.
>     After a deprecation cycle\, its magic was removed in Perl v5\.10\.0\.
>     Using it now triggers a warning&#8203;: C\<$\* is no longer supported>\.
>     You should use the C\</s> and C\</m> regexp modifiers instead\.
> 
> This is true\. In that currently a warning is issued&#8203;:
> 
> $ \./perl \-e '$a = $\*'
> $\* is no longer supported at \-e line 1\.
> 
> However\, there's no such warning for other things in that typeglob\.
> 
> $ \./perl \-Ilib \-e '@&#8203;\* = \( "a"\, "b" \); print "\\$\*\[1\] is $\*\[1\]\\n"'
> $\*\[1\] is b
> $ \./perl \-Ilib \-e '%\* = \( "a"\, "b" \); print "\\$\*\{a\} is $\*\{a\}\\n"'
> $\*\{a\} is b
> $ \./perl \-Ilib \-e '\*\* = sub \{ print "Hello world\\n" \}; &\*'
> Hello world
> 
> 
> I think that we should deprecate the use of \*all\* of these\.

\+1\!

> With the intent of freeing up the parser so that things like
> 
>    @&#8203;\*array
> 
> can become meaningful in the near future\. \(Slurpy arrays and slurpy hashes
> would be particularly useful for signatures\)
> 
> 
> I've tried to use csearch to find instances of use of these constructions on
> CPAN\, but it's very hard to avoid getting a heck of a lot of output from
> unrelated things\. As best I could work out \(using a more restricted search
> term\)\, they aren't used\. I think that the biggest user may well be JAPHs and
> obfuscations relying on this&#8203;:
> 
> $ \./perl \-Ilib \-le 'print \*\*\*\*\*\*'
> 1
> 
> Nicholas Clark
> 


\-\- 
H\.Merijn Brand  http&#8203;://tux\.nl   Perl Monger  http&#8203;://amsterdam\.pm\.org/
using perl5\.00307 \.\. 5\.17   porting perl5 on HP\-UX\, AIX\, and openSUSE
http&#8203;://mirrors\.develooper\.com/hpux/        http&#8203;://www\.test\-smoke\.org/
http&#8203;://qa\.perl\.org   http&#8203;://www\.goldmark\.org/jeff/stupid\-disclaimers/

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @xdg

On Tue\, Feb 26\, 2013 at 9&#8203;:33 AM\, H\.Merijn Brand \<h\.m\.brand@&#8203;xs4all\.nl> wrote&#8203;:
>> I think that we should deprecate the use of \*all\* of these\.
>
> \+1\!

\+1


\-\- 
David Golden \<xdg@&#8203;xdg\.me>
Take back your inbox\! → http&#8203;://www\.bunchmail\.com/
Twitter/IRC&#8203;: @&#8203;xdg

**Comment by p5pRT** [comment]
### From @rjbs

\* Nicholas Clark \<perlbug\-followup@&#8203;perl\.org> \[2013\-02\-26T08&#8203;:41&#8203;:35\]
> I think that we should deprecate the use of \*all\* of these\.

As I've said off\-list\, I am in favor\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Feb 26\, 2013 at 10&#8203;:10&#8203;:09AM \-0500\, Ricardo Signes wrote&#8203;:
> \* Nicholas Clark \<perlbug\-followup@&#8203;perl\.org> \[2013\-02\-26T08&#8203;:41&#8203;:35\]
> > I think that we should deprecate the use of \*all\* of these\.
> 
> As I've said off\-list\, I am in favor\.

Should we deprecate them in time for v5\.18\.0?

Nicholas Clark

**Comment by p5pRT** [comment]
### From @tux

On Tue\, 26 Feb 2013 15&#8203;:53&#8203;:16 \+0000\, Nicholas Clark \<nick@&#8203;ccl4\.org>
wrote&#8203;:

> On Tue\, Feb 26\, 2013 at 10&#8203;:10&#8203;:09AM \-0500\, Ricardo Signes wrote&#8203;:
> > \* Nicholas Clark \<perlbug\-followup@&#8203;perl\.org> \[2013\-02\-26T08&#8203;:41&#8203;:35\]
> > > I think that we should deprecate the use of \*all\* of these\.
> > 
> > As I've said off\-list\, I am in favor\.
> 
> Should we deprecate them in time for v5\.18\.0?

YES\!

> Nicholas Clark


\-\- 
H\.Merijn Brand  http&#8203;://tux\.nl   Perl Monger  http&#8203;://amsterdam\.pm\.org/
using perl5\.00307 \.\. 5\.17   porting perl5 on HP\-UX\, AIX\, and openSUSE
http&#8203;://mirrors\.develooper\.com/hpux/        http&#8203;://www\.test\-smoke\.org/
http&#8203;://qa\.perl\.org   http&#8203;://www\.goldmark\.org/jeff/stupid\-disclaimers/

**Comment by p5pRT** [comment]
### From @rurban

On Tue\, Feb 26\, 2013 at 10&#8203;:11 AM\, H\.Merijn Brand \<h\.m\.brand@&#8203;xs4all\.nl> wrote&#8203;:
> On Tue\, 26 Feb 2013 15&#8203;:53&#8203;:16 \+0000\, Nicholas Clark \<nick@&#8203;ccl4\.org>
> wrote&#8203;:
>
>> On Tue\, Feb 26\, 2013 at 10&#8203;:10&#8203;:09AM \-0500\, Ricardo Signes wrote&#8203;:
>> > \* Nicholas Clark \<perlbug\-followup@&#8203;perl\.org> \[2013\-02\-26T08&#8203;:41&#8203;:35\]
>> > > I think that we should deprecate the use of \*all\* of these\.
>> >
>> > As I've said off\-list\, I am in favor\.
>>
>> Should we deprecate them in time for v5\.18\.0?

Sure

\-\- 
Reini Urban
http&#8203;://cpanel\.net/   http&#8203;://www\.perl\-compiler\.org/

**Comment by p5pRT** [comment]
### From @rjbs

\* Reini Urban \<rurban@&#8203;x\-ray\.at> \[2013\-02\-26T11&#8203;:45&#8203;:23\]
> > On Tue\, 26 Feb 2013 15&#8203;:53&#8203;:16 \+0000\, Nicholas Clark \<nick@&#8203;ccl4\.org>
> > wrote&#8203;:
> >>
> >> Should we deprecate them in time for v5\.18\.0?
> 
> Sure

I'm with Reini\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @nwc10

Merged to blead\, so marking as resolved\.

Nicholas Clark

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Mar 21 04&#8203;:18&#8203;:24 2013\, nicholas wrote&#8203;:
> Merged to blead\, so marking as resolved\.

Hold on a moment\.

$ \./perl \-Ilib \-e '\*\*'
\*\* is deprecated\, and will become a syntax error at \-e line 1\.
$ \./perl \-Ilib \-e '\*\{\*\}'
\*\* is deprecated\, and will become a syntax error at \-e line 1\.
$ \./perl \-Ilib \-e '\*\{"\*"\}'
\*\* is deprecated\, and will become a syntax error at \-e line 1\.
$ \./perl \-Ilib \-e '$\_ = "\*"; \*$\_'
\*\* is deprecated\, and will become a syntax error at \-e line 1\.

That is \*not\* good\.  If you want to deprecate \<sigil>\<asterisk>\, then
toke\.c is the place to do it\.

This close to a stable release\, I would suggest reverting it and
reinstating the deprecation after 5\.18\.

\(Yes\, I’ve been quiet lately\, but real life keeps getting in the way\. 
Every time I think I’m going to have time\, it proves to be more elusive
than I imagined\. :\-\)

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Mar 24\, 2013 at 11&#8203;:46&#8203;:58AM \-0700\, Father Chrysostomos via RT wrote&#8203;:
> On Thu Mar 21 04&#8203;:18&#8203;:24 2013\, nicholas wrote&#8203;:
> > Merged to blead\, so marking as resolved\.
> 
> Hold on a moment\.

Sorry\, this reply took more than a moment\.

> $ \./perl \-Ilib \-e '\*\*'
> \*\* is deprecated\, and will become a syntax error at \-e line 1\.
> $ \./perl \-Ilib \-e '\*\{\*\}'
> \*\* is deprecated\, and will become a syntax error at \-e line 1\.
> $ \./perl \-Ilib \-e '\*\{"\*"\}'
> \*\* is deprecated\, and will become a syntax error at \-e line 1\.
> $ \./perl \-Ilib \-e '$\_ = "\*"; \*$\_'
> \*\* is deprecated\, and will become a syntax error at \-e line 1\.
> 
> That is \*not\* good\.  If you want to deprecate \<sigil>\<asterisk>\, then
> toke\.c is the place to do it\.

Yes\, you are right\. \(But there's more where I'm wrong\, below\)

> This close to a stable release\, I would suggest reverting it and
> reinstating the deprecation after 5\.18\.

I have done this all in a smoke\-me branch\.
I think that it can be done in the tokeniser fairly simply in S\_scan\_ident\(\)\.
It requires another parameter to state whether to warn on encountering "\*"\,
which is only true from the 4 points that call it to parse @&#8203; & \* %\.

so&#8203;:

 STATIC char \*
\-S\_scan\_ident\(pTHX\_ char \*s\, const char \*send\, char \*dest\, STRLEN destlen\, I32 ck\_uni\)
\+S\_scan\_ident\(pTHX\_ char \*s\, const char \*send\, char \*dest\, STRLEN destlen\,
\+             I32 ck\_uni\, char warn\_on\_star\)
 \{
&nbsp;     dVAR;
&nbsp;     char \*bracket = NULL;
@&#8203;@&#8203; \-9318\,6 \+9319\,10 @&#8203;@&#8203; S\_scan\_ident\(pTHX\_ char \*s\, const char \*send\, char \*dest\, STRLEN destlen\, I32 ck
&nbsp;     if \(s \< send
&nbsp;         && \(isIDFIRST\_lazy\_if\(s\, is\_utf8\) || VALID\_LEN\_ONE\_IDENT\(s\, is\_utf8\)\)\)
&nbsp;     \{
\+        if \(warn\_on\_star && \!bracket && \*s == '\*'\)
\+            Perl\_ck\_warner\_d\(aTHX\_ packWARN2\(WARN\_DEPRECATED\, WARN\_SYNTAX\)\,
\+                             "%c\* is deprecated\, and will become a syntax error"\,
\+                             warn\_on\_star\);
&nbsp;         if \(is\_utf8\) \{
&nbsp;             const STRLEN skip = UTF8SKIP\(s\);
&nbsp;             STRLEN i;



and 4 call sites like this&#8203;:

@&#8203;@&#8203; \-5647\,7 \+5647\,7 @&#8203;@&#8203; Perl\_yylex\(pTHX\)
 
&nbsp;     case '\*'&#8203;:
&nbsp;     if \(PL\_expect \!= XOPERATOR\) \{
\-        s = scan\_ident\(s\, PL\_bufend\, PL\_tokenbuf\, sizeof PL\_tokenbuf\, TRUE\);
\+        s = scan\_ident\(s\, PL\_bufend\, PL\_tokenbuf\, sizeof PL\_tokenbuf\, TRUE\, '\*'\);
&nbsp;         PL\_expect = XOPERATOR;
&nbsp;         force\_ident\(PL\_tokenbuf\, '\*'\);
&nbsp;         if \(\!\*PL\_tokenbuf\)
@&#8203;@&#8203; \-5681\,7 \+5681\,7 @&#8203;@&#8203; Perl\_yylex\(pTHX\)
&nbsp;     \}
&nbsp;     PL\_tokenbuf\[0\] = '%';
&nbsp;     s = scan\_ident\(s\, PL\_bufend\, PL\_tokenbuf \+ 1\,
\-        sizeof PL\_tokenbuf \- 1\, FALSE\);
\+                       sizeof PL\_tokenbuf \- 1\, FALSE\, '%'\);
&nbsp;     if \(\!PL\_tokenbuf\[1\]\) \{
&nbsp;         PREREF\('%'\);
&nbsp;     \}
@&#8203;@&#8203; \-6175\,7 \+6175\,7 @&#8203;@&#8203; Perl\_yylex\(pTHX\)
 
&nbsp;     PL\_tokenbuf\[0\] = '&';
&nbsp;     s = scan\_ident\(s \- 1\, PL\_bufend\, PL\_tokenbuf \+ 1\,
\-               sizeof PL\_tokenbuf \- 1\, TRUE\);
\+               sizeof PL\_tokenbuf \- 1\, TRUE\, '&'\);
&nbsp;     if \(PL\_tokenbuf\[1\]\) \{
&nbsp;         PL\_expect = XOPERATOR;
&nbsp;         force\_ident\_maybe\_lex\('&'\);
@&#8203;@&#8203; \-6539\,7 \+6539\,7 @&#8203;@&#8203; Perl\_yylex\(pTHX\)
&nbsp;     if \(PL\_expect == XOPERATOR\)
&nbsp;         no\_op\("Array"\, s\);
&nbsp;     PL\_tokenbuf\[0\] = '@&#8203;';
\-    s = scan\_ident\(s\, PL\_bufend\, PL\_tokenbuf \+ 1\, sizeof PL\_tokenbuf \- 1\, FALSE\);
\+    s = scan\_ident\(s\, PL\_bufend\, PL\_tokenbuf \+ 1\, sizeof PL\_tokenbuf \- 1\, FALSE\, '@&#8203;'\);
&nbsp;     if \(\!PL\_tokenbuf\[1\]\) \{
&nbsp;         PREREF\('@&#8203;'\);
&nbsp;     \}



However\, I think that I'm wrong on two further levels\. I might be expressing
these two in the wrong order&#8203;:



Firstly the intent was to enable syntax of the form @&#8203;\*foo and %\*bar\.


Having @&#8203;\*foo and %\*bar would seem to imply that one can't also have @&#8203;\* or %\*\.

What hadn't sunk is is that we have both $\# and $\#foo \(and $\#$foo\)\, and there
doesn't seem to be a parsing problem with this\. There \*is\* some special casing
for which punctuation vars $\#\.\.\. will work on\, notably only @&#8203;\+\, @&#8203;\- and @&#8203;@&#8203;&#8203;:

&nbsp;    if \(s\[1\] == '\#' && \(isIDFIRST\_lazy\_if\(s\+2\,UTF\) || strchr\("\{$&#8203;:\+\-@&#8203;"\, s\[2\]\)\)\) \{
&nbsp;        PL\_tokenbuf\[0\] = '@&#8203;';


and\, unlike most of Perl 5\, recognising $\#\.\.\. is space sensitive&#8203;:

$ perl \-le '$foo = \[1\.\.3\]; $\# = \\\*STDERR; print $\#\{$foo\}'
$\# is no longer supported at \-e line 1\.
2
$ perl \-le '$foo = "bar"; %\# = \(bar => "Baz"\); $\# = \\\*STDERR; print $\# \{$foo\}'
Baz

\(in the latter\, $\# \{$foo\} is a hash lookup for key $foo of hash %\#\)

but it generally works without surprising anyone\.

As best I can figure out\, one could add @&#8203;\*foo\, &\*foo\, \*\*foo\, %\*foo $\*foo
without removing anything or breaking any sane code on CPAN\. The only code
which I think would change behaviour is that is either using $\* as the
variable for a file handle passed to print \(anything else?\)\, or code which
would parse $\*\+=1 as $\*\+ = 1 instead of $\* \+= 1\, or code which is making
array slices on %\*\. \(Attached are corner cases for $\# parsing\, with
equivalent examples showing that currently C\<$\*> and C\<$\* > are both parsed
the same as C\<$\# >\.\)


So I think that the right thing to do is not to blanket deprecate parsing @&#8203;\*
&\* \*\* %\* and $\*\, but instead change the parser to warn or deprecate on the
specific ambiguous constructions\. Which means that the "new" "needed"
constructions need to come first\. Or at least some idea of them\.

But\, I think I'm wrong again\.


Secondly\, the specific intent was to have consistent "slurpy" syntax for
subroutine signatures\. Consistent with Perl 6\, and consistent between
the Perl 5 signature and regular Perl 5 code\.

And I got this wrong\. In that\, Perl 6 does have @&#8203;\*foo\. But that's a dynamic
scoping lookup\. The slurpy syntax is \*@&#8203;foo\. \(And \*%foo\, and \*$foo\)
http&#8203;://perlcabal\.org/syn/S06\.html\#List\_parameters

For which we don't need to worry about the various sigils used with the \*\*
typeglob at all\. We need to consider how the parser deals with the 3
typeglobs \*@&#8203;\, \*% and \*$\. And based on how $\# and $\#foo are handled\, I think
that everything that is wanted for "new" syntax is currently a syntax error\.
Or\, if not\, all that is currently legal syntax is incredibly obscure corner
cases\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

$ perl \-le '$foo = \[1\.\.3\]; $\# = \\\*STDERR; print $\#\{$foo\}'
$\# is no longer supported at \-e line 1\.
2
$ perl \-le '$foo = "bar"; %\# = \(bar => "Baz"\); $\# = \\\*STDERR; print $\# \{$foo\}'
Baz
$ perl \-le '$foo = "bar"; %\* = \(bar => "Baz"\); $\* = \\\*STDERR; print $\* \{$foo\}'
Baz
$ perl \-le '$foo = "bar"; %\* = \(bar => "Baz"\); $\* = \\\*STDERR; print $\*\{$foo\}'
Baz

$ perl \-le '@&#8203;foo = \(1\.\.3\); $\# = \\\*STDERR; print $\#\{foo\}'
$\# is no longer supported at \-e line 1\.
2
$ perl \-le '%\# = \(foo => "bar"\); $\# = \\\*STDERR; print $\# \{foo\}'
bar
$ perl \-le '%\* = \(foo => "bar"\); $\* = \\\*STDERR; print $\*\{foo\}'
bar
$ perl \-le '%\* = \(foo => "bar"\); $\* = \\\*STDERR; print $\* \{foo\}'
bar

$ perl \-le '$foo = \[1\.\.3\]; $\# = \\\*STDERR; print $\#$foo'$\# is no longer supported at \-e line 1\.
2
$ perl \-le '$foo = \[1\.\.3\]; $\# = \\\*STDERR; print $\# $foo'
$\# is no longer supported at \-e line 1\.
ARRAY\(0x100803ed0\)
$ perl \-le '$foo = \[1\.\.3\]; $\* = \\\*STDERR; print $\*$foo'
$\* is no longer supported at \-e line 1\.
Scalar found where operator expected at \-e line 1\, near "$\*$foo"
&nbsp;        \(Missing operator before $foo?\)
ARRAY\(0x100803ed0\)
$ perl \-le '$foo = \[1\.\.3\]; $\* = \\\*STDERR; print $\* $foo'
$\* is no longer supported at \-e line 1\.
ARRAY\(0x100803ed0\)

$ \./perl \-lwe '@&#8203;foo = \(1\.\.3\); $\# = \\\*STDERR; print $\#&#8203;::foo'
$\# is no longer supported at \-e line 1\.
2
$ \./perl \-lwe '@&#8203;foo = \(1\.\.3\); $\# = \\\*STDERR; print $\# :&#8203;:foo'
$\# is no longer supported at \-e line 1\.
$\# is no longer supported at \-e line 1\.
Bareword found where operator expected at \-e line 1\, near "$\# :&#8203;:foo"
&nbsp;        \(Missing operator before :&#8203;:foo?\)
:&#8203;:foo
$ \./perl \-lwe '@&#8203;foo = \(1\.\.3\); $\* = \\\*STDERR; print $\*&#8203;::foo'
$\* is no longer supported at \-e line 1\.
$\* is no longer supported at \-e line 1\.
Bareword found where operator expected at \-e line 1\, near "$\*&#8203;::foo"
&nbsp;        \(Missing operator before :&#8203;:foo?\)
Operator or semicolon missing before \*&#8203;::foo at \-e line 1\.
Ambiguous use of \* resolved as operator \* at \-e line 1\.
:&#8203;:foo
$ \./perl \-lwe '@&#8203;foo = \(1\.\.3\); $\* = \\\*STDERR; print $\* :&#8203;:foo'
$\* is no longer supported at \-e line 1\.
$\* is no longer supported at \-e line 1\.
Bareword found where operator expected at \-e line 1\, near "$\* :&#8203;:foo"
&nbsp;        \(Missing operator before :&#8203;:foo?\)
:&#8203;:foo

$ perl \-MO=Deparse \-le '$\#\+=1'
BEGIN \{ $/ = "\\n"; $\\ = "\\n"; \}
$\#\+ = 1;
\-e syntax OK
$ perl \-MO=Deparse \-le '$\# \+=1'
$\# is no longer supported at \-e line 1\.
BEGIN \{ $/ = "\\n"; $\\ = "\\n"; \}
$\# \+= 1;
\-e syntax OK
$ perl \-MO=Deparse \-le '$\*\+=1'
$\* is no longer supported at \-e line 1\.
BEGIN \{ $/ = "\\n"; $\\ = "\\n"; \}
$\* \+= 1;
\-e syntax OK
$ perl \-MO=Deparse \-le '$\* \+=1'
$\* is no longer supported at \-e line 1\.
BEGIN \{ $/ = "\\n"; $\\ = "\\n"; \}
$\* \+= 1;
\-e syntax OK

\# $\#\- is analagous to $\#\+\, and the assignments are run time errors

\# Parsing of @&#8203;@&#8203; and $\#@&#8203; was added by commit c0b977fd793319b0 in May 2009&#8203;:

perl5\.14\.0 \-le '@&#8203;@&#8203; = \(1\.\.7\); @&#8203;; = \(1\.\.3\); $\# = \\\*STDERR; print $\#@&#8203;;'
$\# is no longer supported at \-e line 1\.
6
perl5\.14\.0 \-le '@&#8203;@&#8203; = \(1\.\.7\); @&#8203;; = \(1\.\.3\); $\# = \\\*STDERR; print $\# @&#8203;;'
$\# is no longer supported at \-e line 1\.
123
perl5\.14\.0 \-le '@&#8203;@&#8203; = \(1\.\.7\); @&#8203;; = \(1\.\.3\); $\* = \\\*STDERR; print $\*@&#8203;;'
$\* is no longer supported at \-e line 1\.
Array found where operator expected at \-e line 1\, at end of line
&nbsp;        \(Missing operator before ?\)
123
perl5\.14\.0 \-le '@&#8203;@&#8203; = \(1\.\.7\); @&#8203;; = \(1\.\.3\); $\* = \\\*STDERR; print $\* @&#8203;;'
$\* is no longer supported at \-e line 1\.
123

**Comment by p5pRT** [comment]
### From @rjbs

From IRC\, lest it be lost&#8203;:

\<@&#8203;rjbs> Nicholas&#8203;: You asked whether we were okay to revert the imperfect 
deprecation\.  I said\, "I thought you already had\!"
\<@&#8203;rjbs> Nicholas&#8203;: \.\.\.but I think subsequent to that\, it has not actually
been done\, and needs to be\.  Is that correct?

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Apr 30\, 2013 at 04&#8203;:14&#8203;:43PM \-0700\, Ricardo SIGNES via RT wrote&#8203;:
> >From IRC\, lest it be lost&#8203;:
> 
> \<@&#8203;rjbs> Nicholas&#8203;: You asked whether we were okay to revert the imperfect 
> deprecation\.  I said\, "I thought you already had\!"
> \<@&#8203;rjbs> Nicholas&#8203;: \.\.\.but I think subsequent to that\, it has not actually
> been done\, and needs to be\.  Is that correct?

I've pushed the reversion to blead\. To avoid it getting lost\, I've attached
the other commit from that branch here before deleting it\.

There are also two further commits with documentation and tests to apply
after the code freeze in the branch smoke\-me/nicholas/S\_scan\_ident

I keep thinking that there is something that I've missed here\. Hence the
delay\. But if there is\, it hasn't come back to me after several weeks\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

<details><summary>0001-Add-a-deprecation-warning-when-parsing-and.patch</summary>

```diff
From e14d8621b18087e5a3854be4df8aa23b01b74dea Mon Sep 17 00:00:00 2001
From: Nicholas Clark <nick@ccl4.org>
Date: Tue, 9 Apr 2013 15:54:25 +0200
Subject: [PATCH] Add a deprecation warning when parsing @*, %*, &* and **.

Forms such as @{*}, @{"*"} are not deprecated, nor are runtime references
to the variables, such as symbolic references or symbol table manipulation.
---
 embed.fnc         |    2 +-
 embed.h           |    2 +-
 pod/perldiag.pod  |    7 +++++++
 proto.h           |    2 +-
 t/lib/warnings/gv |   34 ++++++++++++++++++++++++++++++++++
 toke.c            |   23 ++++++++++++++---------
 6 files changed, 58 insertions(+), 12 deletions(-)

diff --git a/embed.fnc b/embed.fnc
index ecdde73..9e34ccf 100644
--- a/embed.fnc
+++ b/embed.fnc
@@ -2199,7 +2199,7 @@ iR	|SV*	|get_and_check_backslash_N_name|NN const char* s \
 sR	|char*	|scan_formline	|NN char *s
 sR	|char*	|scan_heredoc	|NN char *s
 s	|char*	|scan_ident	|NN char *s|NN const char *send|NN char *dest \
-				|STRLEN destlen|I32 ck_uni
+				|STRLEN destlen|I32 ck_uni|char warn_on_star
 sR	|char*	|scan_inputsymbol|NN char *start
 sR	|char*	|scan_pat	|NN char *start|I32 type
 sR	|char*	|scan_str	|NN char *start|int keep_quoted \
diff --git a/embed.h b/embed.h
index 96309b2..a4e943e 100644
--- a/embed.h
+++ b/embed.h
@@ -1615,7 +1615,7 @@
 #define scan_const(a)		S_scan_const(aTHX_ a)
 #define scan_formline(a)	S_scan_formline(aTHX_ a)
 #define scan_heredoc(a)		S_scan_heredoc(aTHX_ a)
-#define scan_ident(a,b,c,d,e)	S_scan_ident(aTHX_ a,b,c,d,e)
+#define scan_ident(a,b,c,d,e,f)	S_scan_ident(aTHX_ a,b,c,d,e,f)
 #define scan_inputsymbol(a)	S_scan_inputsymbol(aTHX_ a)
 #define scan_pat(a,b)		S_scan_pat(aTHX_ a,b)
 #define scan_str(a,b,c,d,e)	S_scan_str(aTHX_ a,b,c,d,e)
diff --git a/pod/perldiag.pod b/pod/perldiag.pod
index 3d8212e..8abdd09 100644
--- a/pod/perldiag.pod
+++ b/pod/perldiag.pod
@@ -2642,6 +2642,13 @@ modifiers.  You can enable C</m> for a lexical scope (even a whole file)
 with C<use re '/m'>.  (In older versions: when C<$*> was set to a true value
 then all regular expressions behaved as if they were written using C</m>.)
 
+=item %c* is deprecated, and will become a syntax error
+
+(D deprecated, syntax) The punctuation variables C<@*>, C<&*>, C<**> and
+C<%*> will be removed soon.  In future such code will fail to compile with a
+syntax error.  Removing these variables along with C<$*> will permit future
+syntax additions.
+
 =item $# is no longer supported
 
 (D deprecated, syntax) The special variable C<$#>, deprecated in older
diff --git a/proto.h b/proto.h
index 59ecbc6..0795918 100644
--- a/proto.h
+++ b/proto.h
@@ -7300,7 +7300,7 @@ STATIC char*	S_scan_heredoc(pTHX_ char *s)
 #define PERL_ARGS_ASSERT_SCAN_HEREDOC	\
 	assert(s)
 
-STATIC char*	S_scan_ident(pTHX_ char *s, const char *send, char *dest, STRLEN destlen, I32 ck_uni)
+STATIC char*	S_scan_ident(pTHX_ char *s, const char *send, char *dest, STRLEN destlen, I32 ck_uni, char warn_on_star)
 			__attribute__nonnull__(pTHX_1)
 			__attribute__nonnull__(pTHX_2)
 			__attribute__nonnull__(pTHX_3);
diff --git a/t/lib/warnings/gv b/t/lib/warnings/gv
index 6b022e1..c28dfae 100644
--- a/t/lib/warnings/gv
+++ b/t/lib/warnings/gv
@@ -100,6 +100,7 @@ $* is no longer supported at - line 7.
 $a = $#;
 $a = $*;
 EXPECT
+@* is deprecated, and will become a syntax error at - line 3.
 $# is no longer supported at - line 4.
 $* is no longer supported at - line 5.
 ########
@@ -111,6 +112,39 @@ $a = $*;
 EXPECT
 $# is no longer supported at - line 2.
 $* is no longer supported at - line 3.
+@* is deprecated, and will become a syntax error at - line 5.
+########
+# gv.c
+$a = \@*;
+$a = \&*;
+$a = \**;
+$a = \%*;
+no warnings 'deprecated' ;
+$a = \@*;
+$a = \&*;
+$a = \**;
+$a = \%*;
+EXPECT
+@* is deprecated, and will become a syntax error at - line 2.
+&* is deprecated, and will become a syntax error at - line 3.
+** is deprecated, and will become a syntax error at - line 4.
+%* is deprecated, and will become a syntax error at - line 5.
+########
+# gv.c
+# None of these should warn:
+$a = \@{*};
+$a = \&{*};
+$a = \*{*};
+$a = \%{*};
+$a = \@{"*"};
+$a = \&{"*"};
+$a = \*{"*"};
+$a = \%{"*"};
+$_ = "*";
+$a = \@$_;
+$a = \&$_;
+$a = \*$_;
+$a = \%$_;
 ########
 # gv.c
 use warnings 'syntax' ;
diff --git a/toke.c b/toke.c
index d7b95ed..e6cb503 100644
--- a/toke.c
+++ b/toke.c
@@ -3870,7 +3870,7 @@ S_intuit_more(pTHX_ char *s)
 		weight -= seen[un_char] * 10;
 		if (isWORDCHAR_lazy_if(s+1,UTF)) {
 		    int len;
-		    scan_ident(s, send, tmpbuf, sizeof tmpbuf, FALSE);
+		    scan_ident(s, send, tmpbuf, sizeof tmpbuf, FALSE, 0);
 		    len = (int)strlen(tmpbuf);
 		    if (len > 1 && gv_fetchpvn_flags(tmpbuf, len,
                                                     UTF ? SVf_UTF8 : 0, SVt_PV))
@@ -5656,7 +5656,7 @@ Perl_yylex(pTHX)
 
     case '*':
 	if (PL_expect != XOPERATOR) {
-	    s = scan_ident(s, PL_bufend, PL_tokenbuf, sizeof PL_tokenbuf, TRUE);
+	    s = scan_ident(s, PL_bufend, PL_tokenbuf, sizeof PL_tokenbuf, TRUE, '*');
 	    PL_expect = XOPERATOR;
 	    force_ident(PL_tokenbuf, '*');
 	    if (!*PL_tokenbuf)
@@ -5690,7 +5690,7 @@ Perl_yylex(pTHX)
 	}
 	PL_tokenbuf[0] = '%';
 	s = scan_ident(s, PL_bufend, PL_tokenbuf + 1,
-		sizeof PL_tokenbuf - 1, FALSE);
+                       sizeof PL_tokenbuf - 1, FALSE, '%');
 	if (!PL_tokenbuf[1]) {
 	    PREREF('%');
 	}
@@ -6184,7 +6184,7 @@ Perl_yylex(pTHX)
 
 	PL_tokenbuf[0] = '&';
 	s = scan_ident(s - 1, PL_bufend, PL_tokenbuf + 1,
-		       sizeof PL_tokenbuf - 1, TRUE);
+		       sizeof PL_tokenbuf - 1, TRUE, '&');
 	if (PL_tokenbuf[1]) {
 	    PL_expect = XOPERATOR;
 	    force_ident_maybe_lex('&');
@@ -6417,7 +6417,7 @@ Perl_yylex(pTHX)
 	if (s[1] == '#' && (isIDFIRST_lazy_if(s+2,UTF) || strchr("{$:+-@", s[2]))) {
 	    PL_tokenbuf[0] = '@';
 	    s = scan_ident(s + 1, PL_bufend, PL_tokenbuf + 1,
-			   sizeof PL_tokenbuf - 1, FALSE);
+			   sizeof PL_tokenbuf - 1, FALSE, 0);
 	    if (PL_expect == XOPERATOR)
 		no_op("Array length", s);
 	    if (!PL_tokenbuf[1])
@@ -6429,7 +6429,7 @@ Perl_yylex(pTHX)
 
 	PL_tokenbuf[0] = '$';
 	s = scan_ident(s, PL_bufend, PL_tokenbuf + 1,
-		       sizeof PL_tokenbuf - 1, FALSE);
+		       sizeof PL_tokenbuf - 1, FALSE, 0);
 	if (PL_expect == XOPERATOR)
 	    no_op("Scalar", s);
 	if (!PL_tokenbuf[1]) {
@@ -6548,7 +6548,7 @@ Perl_yylex(pTHX)
 	if (PL_expect == XOPERATOR)
 	    no_op("Array", s);
 	PL_tokenbuf[0] = '@';
-	s = scan_ident(s, PL_bufend, PL_tokenbuf + 1, sizeof PL_tokenbuf - 1, FALSE);
+	s = scan_ident(s, PL_bufend, PL_tokenbuf + 1, sizeof PL_tokenbuf - 1, FALSE, '@');
 	if (!PL_tokenbuf[1]) {
 	    PREREF('@');
 	}
@@ -7817,7 +7817,7 @@ Perl_yylex(pTHX)
 		p = PEEKSPACE(p);
 		if (isIDFIRST_lazy_if(p,UTF)) {
 		    p = scan_ident(p, PL_bufend,
-			PL_tokenbuf, sizeof PL_tokenbuf, TRUE);
+                                   PL_tokenbuf, sizeof PL_tokenbuf, TRUE, 0);
 		    p = PEEKSPACE(p);
 		}
 		if (*p != '$')
@@ -9270,7 +9270,8 @@ S_scan_word(pTHX_ char *s, char *dest, STRLEN destlen, int allow_package, STRLEN
 }
 
 STATIC char *
-S_scan_ident(pTHX_ char *s, const char *send, char *dest, STRLEN destlen, I32 ck_uni)
+S_scan_ident(pTHX_ char *s, const char *send, char *dest, STRLEN destlen,
+             I32 ck_uni, char warn_on_star)
 {
     dVAR;
     char *bracket = NULL;
@@ -9329,6 +9330,10 @@ S_scan_ident(pTHX_ char *s, const char *send, char *dest, STRLEN destlen, I32 ck
     if (s < send
         && (isIDFIRST_lazy_if(s, is_utf8) || VALID_LEN_ONE_IDENT(s, is_utf8)))
     {
+        if (warn_on_star && !bracket && *s == '*')
+            Perl_ck_warner_d(aTHX_ packWARN2(WARN_DEPRECATED, WARN_SYNTAX),
+                             "%c* is deprecated, and will become a syntax error",
+                             warn_on_star);
         if (is_utf8) {
             const STRLEN skip = UTF8SKIP(s);
             STRLEN i;
-- 
1.7.2.5


```
</details>

**Comment by p5pRT** [comment]
### From @rjbs

Thanks\, unblocking\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Apr 09 07&#8203;:33&#8203;:10 2013\, nicholas wrote&#8203;:
> And I got this wrong\. In that\, Perl 6 does have @&#8203;\*foo\. But that's a
> dynamic
> scoping lookup\. The slurpy syntax is \*@&#8203;foo\. \(And \*%foo\, and \*$foo\)
> http&#8203;://perlcabal\.org/syn/S06\.html\#List\_parameters
> 
> For which we don't need to worry about the various sigils used with
> the \*\*
> typeglob at all\. We need to consider how the parser deals with the 3
> typeglobs \*@&#8203;\, \*% and \*$\. And based on how $\# and $\#foo are handled\, I
> think
> that everything that is wanted for "new" syntax is currently a syntax
> error\.
> Or\, if not\, all that is currently legal syntax is incredibly obscure
> corner
> cases\.

Not too incredibly&#8203;:

vec $buf\,1\,1\,\*$foo

I take advantage of that comma\-then\-infix trick all the time\.

And when generating reports\, multiplying by an array length \(\*@&#8203;foo\) is
not so uncommon\, either\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Feb 26 05&#8203;:41&#8203;:34 2013\, nicholas wrote&#8203;:
> perlvar says&#8203;:
> 

Dan Collins did some bookkeeping on this RT today\, which pushed it to the top of the sorted\-by\-last\-modification date in the New queue\, which in turn brought it to my attention\.

My recommendation is that we close this ticket and request that anyone who is interested in its subject matter submit a new ticket\.

\* There has been no discussion in the ticket in 3\-1/4 years\.

\* When there was discussion\, there was \(a\) commits\, then reverts; \(b\) disagreement as to how to proceed\.  Consequently\, it's very difficult to understand what the current status of the ticket is\.

I am taking this ticket for the purpose of closing it in 7 days unless someone takes it over and moves the discussion forward\.

Thank you very much\.


\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Sep 30 12&#8203;:10&#8203;:27 2016\, jkeenan wrote&#8203;:
> On Tue Feb 26 05&#8203;:41&#8203;:34 2013\, nicholas wrote&#8203;:
> > perlvar says&#8203;:
> >
> 
> Dan Collins did some bookkeeping on this RT today\, which pushed it to
> the top of the sorted\-by\-last\-modification date in the New queue\,

Oh\, so I’m not the only one doing that\. :\-\)

> which in turn brought it to my attention\.
> 
> My recommendation is that we close this ticket

I agree\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Sep 30 12&#8203;:36&#8203;:39 2016\, sprout wrote&#8203;:
> On Fri Sep 30 12&#8203;:10&#8203;:27 2016\, jkeenan wrote&#8203;:
> > On Tue Feb 26 05&#8203;:41&#8203;:34 2013\, nicholas wrote&#8203;:
> > > perlvar says&#8203;:
> > >
> > 
> > Dan Collins did some bookkeeping on this RT today\, which pushed it to
> > the top of the sorted\-by\-last\-modification date in the New queue\,
> 
> Oh\, so I’m not the only one doing that\. :\-\)
> 
> > which in turn brought it to my attention\.
> > 
> > My recommendation is that we close this ticket
> 
> I agree\.
> 

Closing as Rejected\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'
